src/Adapter/TaxRulesGroup/CommandHandler/UpdateTaxRuleHandler.php

```diff
@@ -0,0 +1,178 @@
+    ) {
+        parent::__construct($validator);
+
+        $this->langId = $context->getLanguage()->id;
+        $this->countryDataProvider = $countryDataProvider;
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @throws CannotUpdateTaxRuleException
+     * @throws CountryNotFoundException
+     */
+    public function handle(UpdateTaxRuleCommand $command): void
+    {
+        try {
+            $taxRule = new TaxRule($command->getTaxRuleId()->getValue());
+        } catch (PrestaShopException $e) {
+            throw new CannotUpdateTaxRuleException([]);
+        }
+
+        $taxRulesGroupId = (int) $taxRule->id_tax_rules_group;
+        $zipCode = $command->getZipCode();
+        $taxId = $command->getTaxId() !== null ? $command->getTaxId()->getValue() : 0;
+        $behavior = $command->getBehaviorId()->getValue();
+        $description = $command->getDescription();
+
+        $taxRule->behavior = $behavior;
+        $taxRule->id_tax = $taxId;
+        $taxRule->description = $description;
+
+        $selectedCountries = $this->getCountryForTaxRule(
+            $this->countryDataProvider,
+            $this->langId,
+            $command->getCountryId()
+        );
+
+        $stateIds = $command->getStateIds();
+
+        $countriesWithErrors = [];
+
+        foreach ($selectedCountries as $countryKey => $country) {
+            $statesWithErrors = [];
+
+            /** @var StateId $stateId */
+            foreach ($stateIds as $stateKey => $stateId) {
+                $error = false;
+                try {
+                    $hasRuleForCountry = $this->assertUniqueBehaviorTaxRuleForCountry(
+                        $taxRulesGroupId,
+                        $country,
+                        $stateId->getValue(),
+                        $taxRule->id
+                    );
+
+                    if ($hasRuleForCountry) {
+                        $error = true;
+                    }
+
+                    if (null !== $command->getZipCode()) {
+                        $isInvalid = $this->assertIsValidZipCode(
+                            $zipCode,
+                            $country
+                        );
+
+                        if ($isInvalid) {
+                            $error = true;
+                        }
+                    }
+
+                    $taxRule->id_country = $country;
+                    $taxRule->id_state = $stateId->getValue();
+                    list($taxRule->zipcode_from, $taxRule->zipcode_to) = $taxRule->breakDownZipCode($zipCode);
+
+                    try {
+                        $this->validateTaxRuleFields($taxRule);
+                    } catch (TaxRuleConstraintException $e) {
+                        $error = true;
+                    }
+
+                    reset($selectedCountries);
+                    reset($stateIds);
+
+                    if ($countryKey === key($selectedCountries) && $stateKey === key($stateIds)) {
```

Comment:
I dont understand what is happening here 🤔 😉

Reply:
In tax rule edit form there is a possibility to change tax rule country from single to All. In this case first tax rule should be updated and the rest should be created with the rest of the countries. At least legacy code intended that behavior but instead in this case updated tax rule country to Zimbabwe 😄

Reply:
That is a funny bug 😄 I reported it here https://github.com/PrestaShop/PrestaShop/issues/15689

We need this explanation in the code. To do so, we have multiple options:
- write it in a comment
- create a dedicated private function to perform this logic, and use the function name to explain what is going on
- create the private function and put the explanation in the phpDoc

I let you choose 😉